README.md
# kannel
Kannel: Open Source WAP and SMS gateway


gateway/ChangeLog
2019-01-21 Alexander Malysh <
    * gw/bb_boxc.c: fixed small mem leak wile initializing of box_allow_ip/box_deny_ip.
      Thanks to Paul Bagyenda for report.

2018-11-22  Stipe Tolj  <stolj at kannel.org>
    * gwlib/date.c: fix case where, struct universaltime element offset_sec may
      be initialized with random value on MacOS X 10.11.x (clang-800.0.42.1), 
      but is then used  in the following up call to date_convert_universal(), so 
      make sure we set this to zero as we're handlign GMT times anyway. NLC.
      
2018-11-03  Stipe Tolj  <stolj at kannel.org>
    * gw/smsbox.c: re-intend smsbox_req_handle(). NLC.
    
2018-10-01  Stipe Tolj  <stolj at kannel.org>
    * gw/bb_boxc.c: use init_smsbox_routes() for initial and reload cases, use
      smsbox_restart() to reload smsbox routing.
    * gw/bb_http.c: interpret bb_graceful_restart() outcome in HTTP admin command
      function httpd_graceful_restart().
    * gw/bearerbox.c: use gw_log_force() for starting/stopping log evetns that
      we want to see in the log, no matter waht log-level is used (forced),
      reload config in bb_graceful_restart() and call the sub-module reload
      functions.
    * gw/bearebox.h: change prototype for smsc2_graceful_restart().
    Thhis patchsets implement the reloading (aka graceful restart) capabilities
    for the MO direction smsbox routing groups.

2018-09-26 Alexander Malysh <amalysh @ kannnel.org>
    * gwlib/date.[ch]: added handling of fraction

2018-09-26 Alexander Malysh <amalysh @ kannnel.org>
    * gwlib/date.[ch]: added handling of timezone

2018-09-25  Stipe Tolj  <stolj at kannel.org>
   * gwlib/http.c: fixing PANIC condition within conn_unregister() in case the
     connection was destroyed already.
     [Msg-Id: <5B964593.1070105@kannel.org>]
     
2018-09-07 Alexander Malysh <amalysh @ kannnel.org>
   * gw/smsc/smsc_http.c, gw/smsc/smsc_http_p.h, gw/smsc/http/*.c: moved kannel
     callbacks to gw/smsc/http/kannel.c. Added gw/smsc/http/*.c to Makefile.am
     and use new gw_dlopen_get_symbol to get callback symbols.

2018-09-07 Alexander Malysh <amalysh @ kannnel.org>
   * gwlib/gw-dlopen.[ch]: added simple dlopen lookup functionality.

2018-09-07  Stipe Tolj  <stolj at kannel.org>
   * configure.in: removing own binary suffix handling, automake provides an own
     construct for this.
     [https://www.gnu.org/software/automake/manual/automake.html#Renaming]
     
2018-09-04  Stipe Tolj  <stolj at kannel.org>
   * test/test_http.c: fix compiler warnings. NLC.
   
2018-09-04  Stipe Tolj  <stolj at kannel.org>
   * test/fakewap.c: cast panic() output to fix compiler warning on 64bit.
  
2018-09-04  Stipe Tolj  <stolj at kannel.org>
   * doc/doxygen/kannel.conf: re-tag the config with the current svn snapshot 
     revision number.
     
2018-08-30  Stipe Tolj  <stolj at kannel.org>
   * test/test_dbpool.c: fix compiler warning/error in case we compile without
     any DB support. NLC.
     
2015-12-14  Stipe Tolj  <stolj at kannel.org>
    * doc/examples/store-redis.conf: add documentation for the 'hash' config
      directive in the 'group = store-db' context.
    * gw/bb_store_redis.c: use redis with binds list and implement an 'hash'
      mode that allows for readable object in the redis store, rather then the
      string blobs as sinle key/value pair.
    * gw/dlr_redis.c: use the binds list values.
    * gwlib/cfg.def: add 'hash' config directive to 'group = store-db' context.
    * gwlib/dbpool_redis.c: implement the usage of redisCommandArgv() with the
      binds values via the list.
    This patchset re-writes some parts of the redis support modules, mainly to
    support the more generic redisCommandArgv() API call that makes things
    more robust internally.

2015-11-16 Alexander Malysh <amalysh at kannel.org>
    * gw/bb_smscconn.c: fixed crash in smsc2_find if id or admin_id is NULL.

2015-10-23 Alexander Malysh <amalysh at kannel.org>
    * gw/smsbox.c: fixed xml post parsing. Check ud type and if not binary
      don't try to hex decode.

2015-10-01  Stipe Tolj  <stolj at kannel.org>
    * gw/sms.[ch]: incubate function sms_charset_processing() into module.
    * gw/smsbox.c: removed charset_processing().
    * gw/smsc/smsc_http.c: fix missing 'charset' processing for type kannel
      when receiving MOs.

2015-09-25  Stipe Tolj  <stolj at kannel.org>
    * utils/mtbatch.c: add option -x to disallow injection of smsbox-id to MT
      messages, in order to let normal smsbox connection handle DLRs.
      
2015-08-04 Alexander Malysh <amalysh at kannel.org>
    * gwlib/conn.c: added support for chained certificates.

2015-07-20  Stipe Tolj  <stolj at kannel.org>
    * gw/dlr.c: add the original msg DLR bitmask to the meta-data of the DLR, 
      allowing smsbox applications to retrieve this information on-the-fly
      without the need to do local storage/retrieval.
    * gw/meta_data.h: define the macros for the DLR bitmask of the orig msg.
    
2015-07-13 Alexander Malysh <amalysh at kannel.org>
    * gw/bb_store.[ch], gw/bb_store_file.c, gw/bb_store_redis.c, gw/bb_store_spool.c:
      Applied patch that fixes store file race condition by status request. Additionally
      this patch removes status from each of store and add it to bb_store to abstract and
      simplify status handling in store.

2015-07-01  Stipe Tolj  <stolj at kannel.org>
    * doc/userguide/userguide.xml: add documentation for escape code %N for DLR
      notification text.
    * gw/bb_alog.c: implement %N escape code for 'access-log-format' pattern.
    
2015-06-24 Alexander Malysh <amalysh at kannel.org>
    * configure.in, configure: fixed library search path order issue on OSX with
      MacPorts installed where iconv library was not found for linking.

2015-06-24 Alexander Malysh <amalysh at kannel.org>
    * utils/mtbatch.1: fixed bad formatting. This fixes #728.
      Thanks to Kalle Niemitalo. 

2015-05-18  Stipe Tolj  <stolj at kannel.org>

gateway/gwlib/date.c
/* ==================================================================== 
 * The Kannel Software License, Version 1.0 
 * 
 * Copyright (c) 2001-2018 Kannel Group  
 * Copyright (c) 1998-2001 WapIT Ltd.   
 * All rights reserved. 
 * 
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions 
 * are met: 
 * 
 * 1. Redistributions of source code must retain the above copyright 
 *    notice, this list of conditions and the following disclaimer. 
 * 
 * 2. Redistributions in binary form must reproduce the above copyright 
 *    notice, this list of conditions and the following disclaimer in 
 *    the documentation and/or other materials provided with the 
 *    distribution. 
 * 
 * 3. The end-user documentation included with the redistribution, 
 *    if any, must include the following acknowledgment: 
 *       "This product includes software developed by the 
 *        Kannel Group (http://www.kannel.org/)." 
 *    Alternately, this acknowledgment may appear in the software itself, 
 *    if and wherever such third-party acknowledgments normally appear. 
 * 
 * 4. The names "Kannel" and "Kannel Group" must not be used to 
 *    endorse or promote products derived from this software without 
 *    prior written permission. For written permission, please  
 *    contact org@kannel.org. 
 * 
 * 5. Products derived from this software may not be called "Kannel", 
 *    nor may "Kannel" appear in their name, without prior written 
 *    permission of the Kannel Group. 
 * 
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED 
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES 
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
 * DISCLAIMED.  IN NO EVENT SHALL THE KANNEL GROUP OR ITS CONTRIBUTORS 
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,  
 * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT  
 * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR  
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,  
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE  
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ==================================================================== 
 * 
 * This software consists of voluntary contributions made by many 
 * individuals on behalf of the Kannel Group.  For more information on  
 * the Kannel Group, please see <http://www.kannel.org/>. 
 * 
 * Portions of this software are based upon software originally written at  
 * WapIT Ltd., Helsinki, Finland for the Kannel project.  
 */ 

/* date.c - utility functions for handling times and dates
 *
 * Richard Braakman
 */

#include <unistd.h>
#include <ctype.h>
#include <string.h>

#include "gwlib.h"

static char const *wkday[7] = {
    "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
};

static char const *monthname[12] = {
    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
};

/* The starting day of each month, if there's not a leap year.
 * January 1 is day 0, December 31 is day 355. */
static int monthstart[12] = {
    0, 31, 59, 90, 120, 151,
    181, 212, 243, 273, 304, 334
};

/* Value in seconds */
#define MINUTE 60
#define HOUR (60 * MINUTE)
#define DAY (24 * HOUR)

Octstr *date_format_http(unsigned long unixtime)
{
    struct tm tm;
    char buffer[30];

    tm = gw_gmtime((time_t) unixtime);

    /* Make sure gmtime gave us a good date.  We check this to
     * protect the sprintf call below, which might overflow its
     * buffer if the field values are bad. */
    if (tm.tm_wday < 0 || tm.tm_wday > 6 ||
        tm.tm_mday < 0 || tm.tm_mday > 31 ||
        tm.tm_mon < 0 || tm.tm_mon > 11 ||
        tm.tm_year < 0 ||
        tm.tm_hour < 0 || tm.tm_hour > 23 ||
        tm.tm_min < 0 || tm.tm_min > 59 ||
        tm.tm_sec < 0 || tm.tm_sec > 61) {
        warning(0, "Bad date for timestamp %lu, cannot format.",
                unixtime);
        return NULL;
    }

    sprintf(buffer, "%s, %02d %s %04d %02d:%02d:%02d GMT",
            wkday[tm.tm_wday], tm.tm_mday, monthname[tm.tm_mon],
            tm.tm_year + 1900, tm.tm_hour, tm.tm_min, tm.tm_sec);

    return octstr_create(buffer);
}

long date_convert_universal(struct universaltime *t)
{
    long date;
    int leapyears;
    long year;

    date = (t->year - 1970) * (365 * DAY);

    /* If we haven't had this year's leap day yet, pretend it's
     * the previous year. */
    year = t->year;
    if (t->month <= 1)
        year--;

    /* Add leap years since 1970.  The magic number 477 is the value
     * this formula would give for 1970 itself.  Notice the extra
     * effort we make to keep it correct for the year 2100. */
    leapyears = (year / 4) - (year / 100) + (year / 400) - 477;
    date += leapyears * DAY;

    date += monthstart[t->month] * DAY;
    date += (t->day - 1) * DAY;
    date += t->hour * HOUR;
    date += t->minute * MINUTE;
    date += t->second;
    date -= t->offset_sec;

    return date;
}

long date_parse_http(Octstr *date)
{
    long pos;
    struct universaltime t;
    Octstr *monthstr = NULL;

    /* First, skip the leading day-of-week string. */
    pos = octstr_search_char(date, ' ', 0);
    if (pos < 0 || pos == octstr_len(date) - 1)
        return -1;
    pos++;  /* Skip the space */

    /* Distinguish between the three acceptable formats */
    if (isdigit(octstr_get_char(date, pos)) &&
        octstr_get_char(date, pos + 2) == ' ') {
        if (octstr_len(date) - pos < (long)strlen("06 Nov 1994 08:49:37 GMT"))
            goto error;
        if (octstr_parse_long(&t.day, date, pos, 10) != pos + 2)
            goto error;
        monthstr = octstr_copy(date, pos + 3, 3);
        if (octstr_parse_long(&t.year, date, pos + 7, 10) != pos + 11)
            goto error;
        if (octstr_parse_long(&t.hour, date, pos + 12, 10) != pos + 14)
            goto error;
        if (octstr_parse_long(&t.minute, date, pos + 15, 10) != pos + 17)
            goto error;
        if (octstr_parse_long(&t.second, date, pos + 18, 10) != pos + 20)
            goto error;
        /* Take the GMT part on faith. */
    } else if (isdigit(octstr_get_char(date, pos)) &&
               octstr_get_char(date, pos + 2) == '-') {
        if (octstr_len(date) - pos < (long)strlen("06-Nov-94 08:49:37 GMT"))
            goto error;
        if (octstr_parse_long(&t.day, date, pos, 10) != pos + 2)
            goto error;
        monthstr = octstr_copy(date, pos + 3, 3);
        if (octstr_parse_long(&t.year, date, pos + 7, 10) != pos + 9)
            goto error;
        if (t.year > 60)
            t.year += 1900;
        else
            t.year += 2000;
        if (octstr_parse_long(&t.hour, date, pos + 10, 10) != pos + 12)
            goto error;
        if (octstr_parse_long(&t.minute, date, pos + 13, 10) != pos + 15)
            goto error;
        if (octstr_parse_long(&t.second, date, pos + 16, 10) != pos + 18)
            goto error;
        /* Take the GMT part on faith. */
    } else {
        if (octstr_len(date) - pos < (long)strlen(" 6 08:49:37 1994"))
            goto error;
        monthstr = octstr_copy(date, pos, 3);
        if (octstr_parse_long(&t.day, date, pos + 4, 10) != pos + 6)
            goto error;
        if (octstr_parse_long(&t.hour, date, pos + 7, 10) != pos + 9)
            goto error;
        if (octstr_parse_long(&t.minute, date, pos + 10, 10) != pos + 12)
            goto error;
        if (octstr_parse_long(&t.second, date, pos + 13, 10) != pos + 15)
            goto error;
        if (octstr_parse_long(&t.year, date, pos + 16, 10) != pos + 20)
            goto error;
    }

    for (t.month = 0; t.month < 12; t.month++) {
        if (octstr_str_compare(monthstr, monthname[t.month]) == 0)
            break;
    }
    if (t.month == 12)
        goto error;

    octstr_destroy(monthstr);
    /* GTM time means offset 0 */
    t.offset_sec = 0;
    return date_convert_universal(&t);

error:
    octstr_destroy(monthstr);
    return -1;
}

int date_parse_iso (struct universaltime *ut, Octstr *os)
{
    int n = 0;
    char *p, *q;

    /* assign defaults */
    ut->month = 0;
    ut->day = 1;
    ut->hour = 0;
    ut->minute = 0;
    ut->second = 0;
    ut->offset_sec = 0;

    p = octstr_get_cstr(os);
    q = p + ((n = octstr_search_char(os, 'T', 0)) >= 0 ? n : octstr_len(os)); /* stop at the end of string or at the time separator */
    if (sscanf(p, "%4ld%n", &ut->year, &n) < 1)
        return -1;
    p += n;

    if (ut->year < 70)
        ut->year += 2000;
    else if (ut->year < 100)
        ut->year += 1900;
     
    while (p < q && !gw_isdigit(*p))
        p++;
    if (sscanf(p, "%2ld%n", &ut->month, &n) < 1)
        return 0;
    p += n;
     
     /* 0-based months */
    if (ut->month > 0)
        ut->month--;
     
    while (p < q && !gw_isdigit(*p))
        p++;     
    if (sscanf(p, "%2ld%n", &ut->day, &n) < 1)
        return 0;
    p += n;

    if (*q == 'T') 
        p = q+1;
    else
        return 0;

    while (*p && !gw_isdigit(*p))
        p++;
    if (sscanf(p, "%2ld%n", &ut->hour, &n) < 1)
        return 0;
    p += n;

    while (*p && !gw_isdigit(*p))
        p++;
    if (sscanf(p, "%2ld%n", &ut->minute, &n) < 1)
        return 0;
    p += n;
     
    while (*p && !gw_isdigit(*p))
        p++;
    if (sscanf(p, "%2ld%n", &ut->second, &n) < 1)
        return 0;
    p += n;

    if (*p == ',') {
        long fract;
        /* fraction */
        p++;
        if (sscanf(p, "%ld%n", &fract, &n) < 1)
            return 0;
        p += n;
    }

    /* check timezone */
    if (*p) {
        long hh = 0, mi = 0;
        char plus = *p;
        p++;

        if (plus == 'Z')
            return 0; /* we are done */

        if (sscanf(p, "%2ld%n", &hh, &n) < 1)
            return -1;
        p += n;

        while (*p && !gw_isdigit(*p))
            p++;
        if (sscanf(p, "%2ld%n", &mi, &n) > 1)
            p += n;

        ut->offset_sec = hh * HOUR + mi * MINUTE;
        if (plus == '-')
            ut->offset_sec = -ut->offset_sec;
    }

    return 0;
}

Octstr* date_create_iso(time_t unixtime) 
{
    struct tm tm;

    tm = gw_gmtime(unixtime);
    
    return octstr_format("%d-%02d-%02dT%02d:%02d:%02dZ", 
        tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday, tm.tm_hour, tm.tm_min, tm.tm_sec);    
}


/* Note that this implementation makes unportable assumptions about time_t. */
long date_universal_now(void)
{
    return (long) time(NULL);
}
